crates/language-server/src/main.rs

```diff
@@ -1,19 +1,48 @@
-mod db;
-mod diagnostics;
-mod goto;
+mod backend;
+mod functionality;
+mod globals;
+mod logger;
 mod server;
-mod state;
 mod util;
-mod workspace;
 
-use db::Jar;
-mod handlers {
-    pub mod notifications;
-    pub mod request;
-}
+use backend::db::Jar;
+use backend::Backend;
+use tracing::Level;
+
+use server::Server;
+
+use crate::logger::{handle_log_messages, setup_logger};
+
+#[tokio_macros::main]
+async fn main() {
+    let stdin = tokio::io::stdin();
+    let stdout = tokio::io::stdout();
+    let rx = setup_logger(Level::INFO).unwrap();
+
+    let (message_senders, message_receivers) = server::setup_message_channels();
+    let (service, socket) =
+        tower_lsp::LspService::build(|client| Server::new(client, message_senders)).finish();
+    let server = service.inner();
+
+    let client = server.client.clone();
+    let mut backend = Backend::new(client);
+
+    // separate runtime for the backend
+    // let backend_runtime = tokio::runtime::Builder::new_multi_thread()
+    //     .worker_threads(4)
+    //     .enable_all()
+    //     .build()
+    //     .unwrap();
 
-use server::run_server;
+    // backend_runtime.spawn(backend.handle_streams());
 
-fn main() {
-    let _ = run_server();
+    tokio::select! {
+        // setup logging
+        _ = handle_log_messages(rx, server.client.clone()) => {},
+        // start the server
+        _ = tower_lsp::Server::new(stdin, stdout, socket)
+            .serve(service) => {}
+        // backend
+        _ = functionality::streams::setup_streams(&mut backend, message_receivers) => {}
+    }
```

Comment:
I don't understand why you need `tokio::select` here.

Is there any reason that you can't do the one below?
```rust
// Prepare channels for `Server` <-> `Backend` communication.
let (tx, rx) = make_channels(...);

let backend = Backend::new(rx, ...);
// Run backend in a child thread.
tokio::spawn(backend.run());

// Make `Server`, `Backend` and the thread for logging is managed by 
// this server via channels for e.g., graceful shutdown.
let server = ...

// Run server.
server.serve(service).await;
```